GillianCore/engine/Abstraction/Unifier.ml

```diff
@@ -955,210 +1022,249 @@ module Make
+                  raise
+                    (Failure
+                      (Format.asprintf
+                          "Unification failure: do not know all ins for %a"
+                          Formula.pp f))
+              | Some pf ->
+                  let discharges_pf =
+                    List.fold_left
+                      (fun ac x -> if ac = Formula.True then x else And (ac, x))
+                      True discharges
+                  in
+                  let discharges_pf =
+                    Reduction.reduce_formula ~unification:true discharges_pf
+                  in
+                  if State.assert_a state [ And (pf, discharges_pf) ] then
+                    USucc astate
+                  else
+                    let vs = State.unfolding_vals state [ pf ] in
+                    UFail [ EAsrt (vs, Not pf, [ [ Pure pf ] ]) ]))
+      | Types les ->
+          let corrections =
+            List.fold_left
+              (fun (ac : Formula.t list) (le, t) ->
+                let v_le = (subst_in_expr_opt astate subst) le in
+                let v_le : vt =
+                  match v_le with
+                  | Some v_le -> v_le
+                  | None -> raise (Failure "DEATH. unify assertion Types")
+                in
+                match State.get_type state v_le with
+                | Some t' -> if t <> t' then False :: ac else ac
+                | None -> Eq (UnOp (TypeOf, Val.to_expr v_le), Lit (Type t)) :: ac)
+              [] les
+          in
 
-        if corrections = [] then USucc astate
-        else
-          let les, _ = List.split les in
-          let les = List.map (subst_in_expr_opt astate subst) les in
-          UFail
-            [
-              EAsrt
-                ( List.map Option.get (List.filter (fun x -> x <> None) les),
-                  Not (Formula.conjunct corrections),
-                  [ [ Pure (Formula.conjunct corrections) ] ] );
-            ]
-    (* LTrue, LFalse, LEmp, LStar *)
-    | _ -> raise (Failure "Illegal Assertion in Unification Plan")
+          if corrections = [] then USucc astate
+          else
+            let les, _ = List.split les in
+            let les = List.map (subst_in_expr_opt astate subst) les in
+            UFail
+              [
+                EAsrt
+                  ( List.map Option.get (List.filter (fun x -> x <> None) les),
+                    Not (Formula.conjunct corrections),
+                    [ [ Pure (Formula.conjunct corrections) ] ] );
+              ]
+      (* LTrue, LFalse, LEmp, LStar *)
+      | _ -> raise (Failure "Illegal Assertion in Unification Plan")
+    )
 
-  and unify_up (s_states : search_state) : up_u_res =
+  and structure_unify_case_reports parent_ids_ref target_case_depth is_new_case astate subst up =
+    let target_case_depth = (if is_new_case then target_case_depth - 1 else target_case_depth) in 
+    let case_depth = List.length !parent_ids_ref in (
+      assert (target_case_depth <= case_depth);
```

Comment:
💀 This fails in most execution
you can try it on DLL_recursive.wisl

Reply:
I believe I've fixed this with 8ecb1cd.